Codechange: Speed up industry generation using industry-type checks.

PeterN opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem

Generating maps with very large amounts of industries and/or with NewGRFs that perform complex position checks can take a long time.

This is partly due to having to iterate the industry pool lots of times to find nearby industries.

<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

In most cases these tests are looking for specific industries types.

So, store a list of industries per industry type. This allows industry generation checks which only consider a specific industry type to check a reduced set of industries, leading to a potential performance increase.

This also removes the need to track industry type counts as well.

<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

More performance can be gained by using other algorithms, but this one is pretty simple at least.

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

SamuXarick commented on src/newgrf_industries.cpp:
```diff
@@ -92,10 +92,11 @@ uint32_t GetIndustryIDAtOffset(TileIndex tile, const Industry *i, uint32_t cur_g
-		if (i->type != type || i == current) continue;
 
-		best_dist = std::min(best_dist, DistanceManhattan(tile, i->location.tile));
+	for (const IndustryID &industry : Industry::industries[type]) {
+		if (industry == current->index) continue;
+
+		best_dist = std::min(best_dist, DistanceManhattan(tile, Industry::Get(industry)->location.tile));
```
```suggestion
	if (type >= NUM_INDUSTRYTYPES) return best_dist;

	for (const IndustryID &industry : Industry::industries[type]) {
		if (industry == current->index) continue;

		best_dist = std::min(best_dist, DistanceManhattan(tile, Industry::Get(industry)->location.tile));
```

`type` can sometimes be passed as 240 or 255.

glx22 replied:
Indeed `type` can be `IT_INVALID` with newGRFs.
https://github.com/OpenTTD/OpenTTD/blob/master/src/newgrf_industries.cpp#L283 and https://github.com/OpenTTD/OpenTTD/blob/master/src/newgrf_industries.cpp#L40-L46

PeterN committed:
Codechange: Speed up industry generation using industry-type checks.

Store a list of industries per industry type. This allows industry generation checks which only consider a specific industry type to check a reduced set of industries, leading to a potential performance increase.

This also removes the need to track industry type counts as well.

SamuXarick commented on src/newgrf_industries.cpp:
```diff
@@ -278,9 +282,12 @@ static uint32_t GetCountAndDistanceOfClosestInstance(uint8_t param_setID, uint8_
 		/* Distance of nearest industry of given type */
-		case 0x64:
+		case 0x64: {
 			if (this->tile == INVALID_TILE) break;
-			return GetClosestIndustry(this->tile, MapNewGRFIndustryType(parameter, indspec->grf_prop.grffile->grfid), this->industry);
+			IndustryType type = MapNewGRFIndustryType(parameter, indspec->grf_prop.grffile->grfid);
+			if (type >= NUM_INDUSTRYTYPES) return UINT32_MAX;
+			return GetClosestIndustry(this->tile, type, this->industry);
+		}
```
Can I ask you a favour? I'd like to cherry pick a commit with only this fix.

2TallTyler commented on src/saveload/industry_sl.cpp:
```diff
@@ -229,7 +228,7 @@ struct INDYChunkHandler : ChunkHandler {
 				LoadMoveAcceptsProduced(i, INDUSTRY_NUM_INPUTS, INDUSTRY_NUM_OUTPUTS);
 			}
-			Industry::IncIndustryTypeCount(i->type);
+			Industry::industries[i->type].push_back(i->index); // Assume savegame indices are sorted.
```
What do they say about assuming? 😛 

Jokes aside, is this a safe assumption here? Honest question, not a critique.

glx22 replied:
Yeah it's safe here, when saving the pool is iterated, and it's ordered.

PeterN replied:
It won't happen with any savegame made by OpenTTD, but the savegame format itself could have out-of-order indices if a file as somehow modified.

2TallTyler reviewed (approved):
#YOLO

(I did review it, just know this PR was a topic of lengthy discussion about how best to do it... and I am fine with your approach here!)
